merge/wipe_fields.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package merge
+
+import (
+	"fmt"
+
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+// WipeManagedFields clears the managedFields for the current manager from any fields that
+// have been/will be removed from the object during preparation.
+// The fields to clear are determined based on the fields that were not added or modified in the preparedObject
+// compared to the liveObject
+func WipeManagedFields(
+	liveManagedFields, newManagedFields fieldpath.ManagedFields,
+	manager string,
+	liveObject, preparedObject *typed.TypedValue,
+) (fieldpath.ManagedFields, error) {
+	newManaged, exists := newManagedFields[manager]
+	if !exists {
+		return newManagedFields, nil
+	}
+
+	liveManaged, exists := liveManagedFields[manager]
+	if !exists {
+		liveManaged = fieldpath.NewVersionedSet(fieldpath.NewSet(), newManaged.APIVersion(), newManaged.Applied())
+	}
+
+	comparisonLivePrepared, err := liveObject.Compare(preparedObject)
+	if err != nil {
+		return nil, fmt.Errorf("comparing live and prepared: %w", err)
+	}
+
+	addedManagedFields := newManaged.Set().Difference(liveManaged.Set())
```

Comment:
Can you try to explain what is the idea behind that technique? :-)

Reply:
In here or in a comment?
The idea is, that in this line we find out which entries we're added by the Update compared to the current managedFields on the liveObject.

In the lines below we remove all entries that would be gone after the object ran through storage (PrepareForCreate/Update), so we end up with only the managedFields entry additions that will be persisted in the end. Then we only write those to the final managedFields.

Reply:
In the top-level comment. I think we need to make it super clear what this does when someone has to modify/and understand what this does, why and how!